### ISSUE TITLE
Incorrect use of Fieldset-legend to group custom radio buttons / checkboxes

### ISSUE THREAD

**Issue by spanchang** [act-rules/act-rules.github.io#906]
Impacts: SC 1.3.1: grouping relation is not exposed as intended.
Test process:
Step 1: Detect use of fieldset with non-empty legend  
- only rendered content (not hidden with display: none) within scope of this rule
- Off-screen legend is also within scope of this rule
Step 1.1: Check if legend has aria-hidden=true
If yes:Exit test,  flag warning: "Legend uses aria-hidden; compare legend text with accessible name for controls"
If no: Continue test
Step 2: Check for elements  with role=checkbox or role=radio without a type attribute
Note: If type=checkbox or type=radio is present, it is perhaps a native HTML INPUT element 
Step 2.1: If type attribute is also present 
Check if it is within a native INPUT element and value of type attribute matches the role
If yes:  exit tes. There is no problem. Unnecessary use of role.  
If no: Exit test. Flag warning message: "INPUT type does not match role for element". 
Step 3: This means  it is a fail condition. Flag a violation.
Remediation:
Use native HTML radio or checkbox controls within a fieldset-legend or use role=group or role=radiogroup with aria-labelledby to  expose grouping relationship.
 
So grouping relationship is not exposed as intended.

**Comment by WilcoFiers** [comment]
I don't think this is a failure of 1.3.1, at least not in every possible scenario. I'm also not aware that ARIA checkboxes / radiobuttons in native fieldsets causes any kind of problem. Can you elaborate?

**Comment by spanchang** [comment]
Wilco,
The focus of this  proposed rule is Web content that  uses custom
controls and fieldset/legend in an attempt to expose grouping
relationship. That is going to fail. The fieldset/legend technique
works only with native HTML form controls.
Native apps  cannot use fieldset/legend markup  and are outside the
scope of this rule.
Here is an example of  the situation to which the rule applies:
<fieldset><legend>Preferred contact hours:</legend>
<div role="radio"  ...> for  say morning hours</div>
<div role="radio"  ...> for  say afternoon hours</div>
<div role="radio"  ...> for  say evening hours</div>
</fieldset>

Clearly, it is the intent to relate the legend text with the custom
radio controls. That will fail. and hence it is a failure of 1.3.1.
I have seen this type of coding on more than one occasion.
Thanks,
Sailesh


On 10/22/19, Wilco Fiers <notifications@github.com> wrote:
> I don't think this is a failure of 1.3.1, at least not in every possible
> scenario. I'm also not aware that ARIA checkboxes / radiobuttons in native
> fieldsets causes any kind of problem. Can you elaborate?
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/act-rules/act-rules.github.io/issues/906#issuecomment-544885922


-- 
Sailesh Panchang
Principal Accessibility Consultant
Deque Systems Inc
381 Elden Street, Suite 2000, Herndon, VA 20170
Mobile: 571-344-1765

**Comment by spanchang** [comment]
Wilco, Also I omitted to say in the last response that  the steps in
the proposed rule, especially step #2 address your concern about
native controls.
Sailesh
On 10/23/19, Sailesh Panchang <sailesh.panchang@deque.com> wrote:
> Wilco,
> The focus of this  proposed rule is Web content that  uses custom
> controls and fieldset/legend in an attempt to expose grouping
> relationship. That is going to fail. The fieldset/legend technique
> works only with native HTML form controls.
> Native apps  cannot use fieldset/legend markup  and are outside the
> scope of this rule.
> Here is an example of  the situation to which the rule applies:
> <fieldset><legend>Preferred contact hours:</legend>
> <div role="radio"  ...> for  say morning hours</div>
> <div role="radio"  ...> for  say afternoon hours</div>
> <div role="radio"  ...> for  say evening hours</div>
> </fieldset>
>
> Clearly, it is the intent to relate the legend text with the custom
> radio controls. That will fail. and hence it is a failure of 1.3.1.
> I have seen this type of coding on more than one occasion.
> Thanks,
> Sailesh
>
>
> On 10/22/19, Wilco Fiers <notifications@github.com> wrote:
>> I don't think this is a failure of 1.3.1, at least not in every possible
>> scenario. I'm also not aware that ARIA checkboxes / radiobuttons in
>> native
>> fieldsets causes any kind of problem. Can you elaborate?
>>
>> --
>> You are receiving this because you authored the thread.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/act-rules/act-rules.github.io/issues/906#issuecomment-544885922
>
>
> --
> Sailesh Panchang
> Principal Accessibility Consultant
> Deque Systems Inc
> 381 Elden Street, Suite 2000, Herndon, VA 20170
> Mobile: 571-344-1765
>


-- 
Sailesh Panchang
Principal Accessibility Consultant
Deque Systems Inc
381 Elden Street, Suite 2000, Herndon, VA 20170
Mobile: 571-344-1765
